### PR TITLE
refactor: add ide extension plugin

### PR DESCRIPTION
**Pull request by felixhaeberle**
this refactors the ide extension functionality out of inlang/core and into a plugin. ref #618

### CODE REVIEW

**Comment by felixhaeberle** [comment]
I've tested this with the sdk-js/adapter-sveltekit/examples/test-app/inlang.config.js with the following config:

```js
import ideExtensionPlugin from "@inlang/ide-extension/plugin"

/**
 * @type { import("@inlang/core/config").DefineConfig }
 */
export async function defineConfig(env) {
	const { default: jsonPlugin } = await env.$import(
		"https://cdn.jsdelivr.net/gh/samuelstroschein/inlang-plugin-json@2/dist/index.js"
	)
	const { default: sdkPlugin } = await env.$import(
		"https://cdn.jsdelivr.net/npm/@inlang/sdk-js@0.0.4/dist/plugin/index.js"
	)
	// const { default: ideExtensionPlugin } = await env.$import(
	// 	"https://cdn.jsdelivr.net/npm/@inlang/ide-extension@x.x.x/dist/plugin/index.js"
	// )

	return {
		referenceLanguage: "en",
		plugins: [
			jsonPlugin({ pathPattern: "./languages/{language}.json" }),
			sdkPlugin({
				languageNegotiation: {
					strategies: [{ type: "localStorage" }, { type: "navigator" }]
				}
			}),
			ideExtensionPlugin({
				messageReferenceMatchers: [
				  async ( /** @type {{ "documentText": string; }} */ args) => {
					const regex = /(?<!\w){?t\(['"](?<messageId>\S+)['"]\)}?/gm;
					const str = args.documentText;
					let match;
					const result = [];
		  
					while ((match = regex.exec(str)) !== null) {
					  const startLine =
						(str.slice(0, Math.max(0, match.index)).match(/\n/g) || [])
						.length + 1;
					  const startPos =
						match.index - str.lastIndexOf("\n", match.index - 1);
					  const endPos =
						match.index +
						match[0].length -
						str.lastIndexOf("\n", match.index + match[0].length - 1);
					  const endLine =
						(
						  str
						  .slice(0, Math.max(0, match.index + match[0].length))
						  .match(/\n/g) || []
						).length + 1;
		  
					  if (match.groups && "messageId" in match.groups) {
						result.push({
						  messageId: match.groups["messageId"],
						  position: {
							start: {
							  line: startLine,
							  character: startPos,
							},
							end: {
							  line: endLine,
							  character: endPos,
							},
						  },
						});
					  }
					}
					return result;
				  },
				],
				extractMessageOptions: [{
					callback: (messageId) => `{i("${messageId}")}`,
				  }
				],
			  })

		],
	}
}

```

**Comment by felixhaeberle** [comment]
@ivanhofer As @samuelstroschein has limited availability, I tagged you as reviewer. 🙏😇

**Comment by felixhaeberle** [comment]
@samuelstroschein @ivanhofer For bringing this to the example repo (and releasing it) – we have to publish a npm package for the ide extension too.

**Comment by ivanhofer** [comment]
@felixhaeberle we need a public url where someone can access the bundled file.
Either build it and commit it in the inlang repository, release it as an npm package, or include it in a repository that can create a "Release" works. I would prefer the `npm` package.
What do you think @samuelstroschein?

File: source-code/ide-extension/src/plugin/schema.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+const ideExtensionSchema = z
+	.object({
+		/**
+		 * Defines matchers for message references inside the code.
+		 *
+		 * @param args represents the data to conduct the search on
+		 * @returns a promise with matched message references
+		 */
+		messageReferenceMatchers: z.array(
```
Comment (ivanhofer) [inline source-code/ide-extension/src/plugin/schema.ts:33]: this seems wrong. Previously it was a function that returned an array

File: source-code/ide-extension/src/plugin/schema.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+		 * An array of VSCode DocumentSelectors.
+		 *
+		 * The document selectors specify for which files/programming languages
+		 * (typescript, svelte, etc.) the extension should be activated.
+		 *
+		 * See https://code.visualstudio.com/api/references/document-selector
+		 */
+		// documentSelectors: DocumentSelector[];
+	})
+	.optional()
```
Comment (ivanhofer) [inline source-code/ide-extension/src/plugin/schema.ts:65]: I would mark it as required, because it doe not make sense to use the plugin and that plugin then does not define the property

**Review by ivanhofer — CHANGES_REQUESTED** [review]
I have fixed a few smaller things. Please take a look at the `zodObject`.
It would also be great to have tests that fail if we pass a wrong config

**Comment by felixhaeberle** [comment]
@ivanhofer Thank you, will do the changes. 🙏

File: source-code/core/src/config/schema.ts [hunk 3]
```diff
@@ -78,54 +78,6 @@ export type InlangConfig = {
...
-				/**
-				 * The position from where to where the reference can be found.
-				 */
-				position: {
-					start: { line: number; character: number }
-					end: { line: number; character: number }
-				}
-			}>
-		>)[]
-
```
[the remark this answers is not in the document]
Comment (ivanhofer) [reply source-code/core/src/config/schema.ts:104]: I never have worked with `z.function` before, but it seems correct. If you add some tests we will know it for sure ;)

**Comment by felixhaeberle** [comment]
@ivanhofer I added tests. but I have to check because currently while writing tests I have some strange type errors.
the zod type doesn't resolve correctly. can you reproduce this?

File: source-code/ide-extension/tsup.config.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import { defineConfig } from "tsup"
```
Comment (samuelstroschein) [inline source-code/ide-extension/tsup.config.ts:1]: I would use pluginBuildConfig from  @inlang/core/plugin to stay compliant.

File: source-code/ide-extension/package.json [hunk 5]
```diff
@@ -19,7 +19,11 @@
 		"Linters",
 		"Visualization"
 	],
-	"main": "./dist/main.cjs",
+	"exports": {
+		".": "./dist/ide/main.cjs",
+		"./plugin": "./dist/plugin/index.js"
+	},
```
Comment (samuelstroschein) [inline source-code/ide-extension/package.json:25]: no exports are needed and might confuse vscode (we are not using npm for plugins and vscode only uses main)

File: source-code/ide-extension/src/plugin/plugin.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import { type IdeExtensionConfig, validateIdeExtensionConfig } from "./schema.js"
+import type { InlangConfig } from "@inlang/core/config"
+import { createPlugin } from "@inlang/core/plugin"
+
+// ------------------------------------------------------------------------------------------------
+
+export const ideExtensionPlugin = createPlugin<IdeExtensionConfig>(({ settings }) => ({
```
Comment (samuelstroschein) [inline source-code/ide-extension/src/plugin/plugin.ts:7]: hey, you told me to not use `pluginConfig` and use `settings` instead :D I would expect `IdeExtensionConfig` to be named `IdeExtensionSettings` ;)

File: source-code/ide-extension/tsup.config.ts (hunk 4 above)
Comment (felixhaeberle) [reply source-code/ide-extension/tsup.config.ts:1]: @samuelstroschein 😄 You missed to export the types correctly from package.json, therefore the import couldn't be found and we didn't use it "in that way already" in the codebase, you imported it from the file directly for the tests. I will fix that tomorrow. 🧼

File: source-code/ide-extension/package.json [hunk 7]
```diff
@@ -28,9 +32,23 @@
 			}
 		]
 	},
+	"typesVersions": {
+		"*": {
+			"plugin": [
+				"./dist/plugin/index.d.ts"
+			]
+		}
+	},
```
[the remark this answers is not in the document]
Comment (felixhaeberle) [reply source-code/ide-extension/package.json:41]: @samuelstroschein How do we publish the ide-extension plugin?

File: source-code/ide-extension/tsup.config.ts (hunk 4 above)
Comment (samuelstroschein) [reply source-code/ide-extension/tsup.config.ts]: The types are exported correctly for node16 module resolution^^

I guess the ide extension (as well as sveltekit) don't use node16 module resolution. All of this module discrepancy is such a pain.

File: source-code/ide-extension/package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (samuelstroschein) [reply source-code/ide-extension/package.json:41]: Just as we publish other plugins, directly via github. 

But, other plugins should build on top of the ide extension plugin. Thus, publishing to npm might be a good idea after all. We can't use `@inlang/ide-extension` as a namespace though because vscode doesn't support `@organization` syntax. 

Might be easier to have two separate package.json files/projects for the ide extension and the ide extension plugin? Should also ease bundling and typing things as the plugin-template can be used. 

```
source-code/ide-extension-plugin [@inlang/plugin-ide-extension] -> this would need to be imported by developers
source-code/ide-extension [vs-code-extension]
```

File: source-code/ide-extension/tsup.config.ts (hunk 4 above)
Comment (ivanhofer) [reply source-code/ide-extension/tsup.config.ts]: I can take a look at this. I think I now have seen all edge cases when it comes to exporting types for all possible formats ^^

**Review by ivanhofer — COMMENTED** [review]
@felixhaeberle I have fixed the failing test for you. Regarding the type issue, I assume you already fixed it yourself, right?
Then the only thing that is missing is to release the plugin as a npm package.

Once the nom package exists, you can fix the failign tests. Currently we can't specify the types, because they don't get exported anywhere.

**Comment by felixhaeberle** [comment]
@ivanhofer thank you!! will have a look on it tomorrow morning ☺️🙏

File: source-code/ide-extension/package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (felixhaeberle) [reply source-code/ide-extension/package.json:41]: @samuelstroschein or even it's own repository like the other plugins / standard lint rules etc? 

then it's really clear that this is some standalone plugin code. 

published under the inlang organization (not my account).

[the remark this answers is not in the document]
Comment (felixhaeberle) [reply source-code/ide-extension/package.json:41]: @samuelstroschein off topic here but regarding the ide extension marketing in the marketplace I would suggest something more common like translation intellisense or something with -"lens", 
-> inlang - i18n translation intellisense, not only "inlang"

we can A/B test this once it's out.

[the remark this answers is not in the document]
Comment (felixhaeberle) [reply source-code/ide-extension/package.json]: @samuelstroschein I have moved it into it's own folder for now. can be further extracted.

File: source-code/ide-extension/src/plugin/plugin.ts (hunk 6 above)
Comment (felixhaeberle) [reply source-code/ide-extension/src/plugin/plugin.ts:6]: @samuelstroschein Done.

**Comment by felixhaeberle** [comment]
@ivanhofer I moved the plugin into it's own folder for an npm release. (@inlang/ide-extension-plugin)
-> source-code/ide-extension-plugin
How can I release it from the command line, or should I use github. I remember we had a command line tool for that?
  
Currently the build / tests are broken by the`sdk-js` workspace, do you know the cause? 
Has it something to do with the missing release?

**Comment by felixhaeberle** [comment]
@ivanhofer After the tests are fixed for the `sdk-js`, this should be merged in order that I can continue preparing the example repo for @samuelstroschein.

**Review by ivanhofer — APPROVED** [review]
I made some changes and think we can release it now
